nova/virt/vmwareapi/driver.py

```diff
@@ -690,6 +697,250 @@ def detach_interface(self, context, instance, vif):
         """Detach an interface from the instance."""
         self._vmops.detach_interface(context, instance, vif)
 
+    @periodic_task.periodic_task(spacing=CONF.vmware.drs_group_sync_interval)
+    def _sync_server_groups(self, compute_manager, context):
+        """Regularly syncs the DB state of server-groups with the cluster
+
+        This function retrieves all groups/rules from the cluster, all
+        instances belonging to this host and all server-groups belonging to
+        these instances. After building a list of expected groups/rules and the
+        expected members from those server-groups, it compares them to the
+        retrieved DRS groups/rules building a list of changes to bring the
+        cluster into expected state. At the end, the changes are applied in
+        batches.
+
+        We only sync DRS groups/rules if they have more then 2 members. We
+        would waste resources on the vCenter otherwise.
+
+        This function is different from other periodic_task's in that it
+        receives the manager as first argument. This comes from the fact, that
+        it's not part of that manager, but instead only registered to it, so it
+        can be called. The argument is unused in this function.
+        """
+        LOG.debug('Starting DRS groups/rules sync-loop.')
+        client_factory = self._session.vim.client.factory
+
+        # retrieve all DRS groups, rules and members
+        cluster_groups = cluster_util.fetch_cluster_groups(
+            self._session, cluster_ref=self._cluster_ref, group_type='vm')
+        cluster_rules = cluster_util.fetch_cluster_rules(
+            self._session, cluster_ref=self._cluster_ref)
+
+        # NOTE: No need to filter the members of DRS groups and rules, because
+        # DRS rules/groups are cluster-local and thus can only contain VMs of
+        # the cluster.
+
+        # filter DRS groups matching to our naming scheme
+        cluster_groups = {name: g for (name, g) in cluster_groups.items()
+                          if name.startswith(constants.DRS_PREFIX)}
+        cluster_rules = {name: r for (name, r) in cluster_rules.items()
+                         if name.startswith(constants.DRS_PREFIX)}
+
+        # get all instance uuids belonging to this host
+        InstanceList = objects.instance.InstanceList
+        instances = InstanceList.get_by_host(context, self._compute_host,
+                                             expected_attrs=[])
+        instances_by_uuid = {i.uuid: i for i in instances}
+        instance_uuids = set(instances_by_uuid)
+
+        # retrieve all InstanceGroups
+        InstanceGroupList = objects.instance_group.InstanceGroupList
+        expected_groups = {}
+        expected_rules = {}
+        ig_list = InstanceGroupList.get_by_instance_uuids(
+            context, set(instances_by_uuid))
```

Comment:
`InstanceGroup` objects are saved in the API DB while hosts are only availabe in the cell's DB. This would mean, we'd have to query the API DB for the list of of server-groups and all their members, then query all cell's DBs for all instances and their host and merge them in Python to filter for the host.

Reply:
I see, then indeed we might as well do it on the host.